dlt/common/plugins/plugins.py

```diff
@@ -0,0 +1,99 @@
+from typing import Type, Union, List, Any, Callable, Iterable
+from dlt.common.typing import TFun
+from dlt.common.typing import TDataItem
+from dlt.common.pipeline import SupportsPipeline
+from dlt.common.configuration.specs import ContainerInjectableContext
+from dlt.common.schema.typing import TSchemaContract
+from dlt.common.configuration.specs import configspec
+from functools import wraps
+from .reference import SupportsCallbackPlugin, Plugin, TSinglePluginArg, TPluginArg, CallbackPlugin
+from dlt.common.configuration.container import Container
+from dlt.common.schema.exceptions import DataValidationError
+
+
+@configspec
+class PluginsContext(ContainerInjectableContext, SupportsCallbackPlugin):
+    def __init__(self) -> None:
+        self._plugins: List[Plugin] = []
+        self._callback_plugins: List[CallbackPlugin] = []
+        self.steps: List[str] = []
+
+    def _resolve_plugin(self, plugin: TSinglePluginArg, pipeline: SupportsPipeline) -> Plugin:
+        if isinstance(plugin, str):
+            pass  # TODO
+        elif isinstance(plugin, type) and issubclass(plugin, Plugin):
+            plugin = plugin()
+        elif isinstance(plugin, Plugin):
+            pass
+        else:
+            raise TypeError(
+                f"Plugin {plugin} is not a subclass of Plugin, nor a Plugin instance, nor a plugin"
+                " name string"
+            )
+        return plugin
+
+    def setup_plugins(self, plugins: TPluginArg, pipeline: SupportsPipeline) -> None:
+        if not plugins:
+            return
+        if not isinstance(plugins, Iterable):
+            plugins = [plugins]
+        for p in plugins:
+            resolved_plugin = self._resolve_plugin(p, pipeline)
+            self._plugins.append(resolved_plugin)
+            if isinstance(resolved_plugin, CallbackPlugin):
+                self._callback_plugins.append(resolved_plugin)
+
+    def on_step_start(self, step: str) -> None:
+        self.steps.append(step)
+        for p in self._plugins:
+            p.step = self.steps[-1]
+            p.on_step_start(step)
+
+    def on_step_end(self, step: str) -> None:
+        self.steps.pop()
+        for p in self._plugins:
+            p.on_step_end(step)
+            p.step = self.steps[-1] if self.steps else None
+
+    #
+    # callback interfaces
+    #
+
+    #
+    # extraction callbacks
+    #
+    def on_extractor_item_written(self, item: TDataItem, **kwargs: Any) -> None:
```

Comment:
hmmmm we have plenty of ways to put any kind of python function at the end of pipeline. so I do not see use for this one.

if you want something really useful: add a callback when we have a state change on the load package in `load.py` (main use case: control retry and fail messages ie by stopping the loading)

Reply:
I removed it for now